server/src/proxyManagement.mjs

```diff
@@ -21,97 +21,114 @@
-  ws = new WebSocket(url)
+
+  outgoingWs.send(command);
+
+  return new Promise(async (res, rej) => {
+    try {
+      const response = await getResponseMessageFromProxy(returnWs);
+      // Clear for next message
+      wsMsgMap.set(returnWs, null);
+      res(response);
+    } catch (err) {
+      rej('Timeout occurred');
+    }
+  });
+}
+
+function setupOutgoingWs(returnWs) {
+  const url = buildWSUrl();
+  const ws = new WebSocket(url);
   try {
     return new Promise((res, rej) => {
       ws.on('open', function open() {
-        console.log("Connection to websocket proxy server established")
-        res(true)
+        console.log('connection established');
+        // add ws connection to map
+        wsMap.set(returnWs, ws);
+        res(ws);
       });
 
       ws.on('close', function close() {
         console.log('disconnected');
-      });
-      
-      ws.on('message', function message(data) {
-        const buf = Buffer.from(data, 'utf8');
-        //send response to callback
-        callback(buf.toString(), returnWS)
+        // remove closed connection from map
+        wsMap.delete(returnWs);
       });
 
       ws.on('error', function message(err) {
-        rej(err)
+        rej(err);
       });
-    })
+    });
   } catch (err) {
-    return err
+    return err;
   }
 }
 
-/* Consume response from server. In case of event, send the event to caller directly.
-In case of response for requested method, store it in responseObject array
-*/
-function actOnResponseObject(data, returnWS) {
-  const response = JSON.parse(data)
-  if(response.id === undefined) {
-    returnWS.send(data)
-  }
-  responseObject[response.id] = data
-}
-
-function sendRequest(command) {
-  if(ws) {
-    ws.send(command)
-    console.log("Request sent to proxy server: ", command);
-  } else {
-    console.log("WS Client not initialized. Unable to send request to proxy server: ", command);
-  }
-}
-
-//Poll for proxy response. Fetch response using requestId. If timedout, terminate the connection
-function getResponseMessageFromProxy(id) {
-  let timeout = 2000
-  let counter = 0
-  let interval = 100
-  return new Promise((resolve, reject) => {
-    var timer = setInterval(function() {
-      if(counter >= timeout) {
-        console.log("response not received for given id: " + id)
-        reject(false)
+function getResponseMessageFromProxy(returnWs) {
+  let timeout = 10000;
+  let counter = 0;
+  let interval = 100;
+  return new Promise((res, rej) => {
+    var timer = setInterval(function () {
+      if (counter >= timeout) {
+        console.log('response not received for given returnWs');
+        rej(false);
       }
-      if(responseObject[id]) {
-        counter = timeout + interval
-        //clear interval if response received for given id.
-        clearInterval(timer)
-        resolve(responseObject[id])
+
+      const returnMsg = wsMsgMap.get(returnWs);
+      if (returnMsg) {
+        counter = timeout + interval;
```

Comment:
Why are we updating the interval here even though we're done?

Reply:
Agreed that this is unnecessary code so I removed line 105